### PR TITLE
native: Uses kevent instead of kevent64

### PR DESCRIPTION
**Pull request by jasonwilliams200OK**
Some operating systems, such as FreeBSD, do not provide kevent64_t but
instead have kevent. This change switches to use kevent for all
platforms.

### CODE REVIEW

**Comment by stephentoub** [comment]
cc: @nguerrera, @sokket, @pgavlin

File: src/Native/System.Native/pal_networking.h [hunk 1]
```diff
@@ -409,7 +409,7 @@ extern "C" Error CreateSocketEventBuffer(int32_t count, SocketEvent** buffer);
 
 extern "C" Error FreeSocketEventBuffer(SocketEvent* buffer);
 
-extern "C" Error TryChangeSocketEventRegistration(int32_t port, int32_t socket, int32_t currentEvents, int32_t newEvents, uintptr_t data);
+extern "C" Error TryChangeSocketEventRegistration(int32_t port, int32_t socket, int32_t currentEvents, int32_t newEvents, int64_t data);
```
Comment (stephentoub) [inline src/Native/System.Native/pal_networking.h]: You're changing the size of a type in a native export, but I don't see an equivalent change on the managed side.

File: src/Native/System.Native/pal_networking.cpp [hunk 2]
```diff
@@ -2257,7 +2275,7 @@ static Error CloseSocketEventPortInner(int32_t port)
     return err == 0 ? PAL_SUCCESS : ConvertErrorPlatformToPal(errno);
 }
 
-static Error TryChangeSocketEventRegistrationInner(int32_t port, int32_t socket, SocketEvents currentEvents, SocketEvents newEvents, uintptr_t data)
+static Error TryChangeSocketEventRegistrationInner(int32_t port, int32_t socket, SocketEvents currentEvents, SocketEvents newEvents, int64_t data)
```
Comment (pgavlin) [inline src/Native/System.Native/pal_networking.cpp]: Please do not change the type of the `data` parameter.

File: src/Native/System.Native/pal_networking.h (hunk 1 above)
Comment (pgavlin) [inline src/Native/System.Native/pal_networking.h]: The size of this parameter shouldn't be changed in any case: these `data` params are intentionally pointer-sized (the values are derived from `GCHandle`s).

File: src/Native/System.Native/pal_networking.cpp [hunk 3]
```diff
@@ -30,6 +30,24 @@
...
+typedef struct kevent kevent64_s;
+
+static void EV_SET64(kevent64_s *kev, uint64_t ident, int16_t filter, uint16_t flags, uint32_t fflags, int64_t data, uint64_t udata, uint64_t ext_0, uint64_t ext_1)
+{
+    (void)ext_0; (void)ext_1; // unused parameters
+    EV_SET(kev, ident, filter, flags, fflags, data, &udata);
+}
+
+static inline int kevent64(int kq, const kevent64_s *changelist, int nchanges, kevent64_s *eventlist, int nevents, unsigned int flags, const struct timespec *timeout) {
+    (void)flags; // unused parameter
```
Comment (nguerrera) [inline src/Native/System.Native/pal_networking.cpp]: nit: brace should go on its own line.

File: src/Native/System.Native/pal_networking.cpp (hunk 2 above)
Comment (jasonwilliams200OK) [inline src/Native/System.Native/pal_networking.cpp]: I wasn't able to test on Linux before. Now I have the access. On FreeBSD and OSX it was compiling.
I will add a #if #else guard, without changing the function signature.

File: src/Native/System.Native/pal_networking.h (hunk 1 above)
Comment (jasonwilliams200OK) [inline src/Native/System.Native/pal_networking.h]: I wasn't able to test on Linux before. Now I have the access. On FreeBSD and OSX it was compiling.
I will add a #if #else guard, without changing the function signature.

File: src/Native/System.Native/pal_networking.cpp [hunk 4]
```diff
@@ -30,6 +30,24 @@
 #include <unistd.h>
 #include <vector>
 
+#if !HAVE_KEVENT64_S
+// On platforms, such as FreeBSD, where kevent64_s
+// is not available, fallback to custom definition
+// with wrapper functions.
+typedef struct kevent kevent64_s;
+
+static void EV_SET64(kevent64_s *kev, uint64_t ident, int16_t filter, uint16_t flags, uint32_t fflags, int64_t data, uint64_t udata, uint64_t ext_0, uint64_t ext_1)
```
Comment (pgavlin) [inline src/Native/System.Native/pal_networking.cpp]: Nit: this could just be a macro definition: `#define EV_SET64(kev, ident, filter, flags, fflags, data, udata, ext_0, ext_1) EV_SET((kev), (ident), (filter), (flags), (fflags), (data), (udata))`

File: src/Native/System.Native/pal_networking.cpp (hunk 3 above)
Comment (jasonwilliams200OK) [inline src/Native/System.Native/pal_networking.cpp]: Ouch! Will fix that in a bit. :smile:

File: src/Native/System.Native/pal_networking.cpp [hunk 5]
```diff
@@ -30,6 +30,24 @@
...
+{
+    (void)ext_0; (void)ext_1; // unused parameters
+    EV_SET(kev, ident, filter, flags, fflags, data, &udata);
```
Comment (pgavlin) [inline src/Native/System.Native/pal_networking.cpp]: Is the `&udata` intentional? Seems like it should be `reinterpret_cast<void*>(udata)`.

File: src/Native/System.Native/pal_networking.cpp (hunk 2 above)
Comment (pgavlin) [inline src/Native/System.Native/pal_networking.cpp]: What do you need the `#if/#else` guard for? It's not clear to me why any change would be necessary here.

File: src/Native/System.Native/pal_networking.cpp (hunk 4 above)
Comment (jasonwilliams200OK) [inline src/Native/System.Native/pal_networking.cpp]: Nice idea. I have changed added the macro.

File: src/Native/System.Native/pal_networking.cpp (hunk 5 above)
Comment (jasonwilliams200OK) [inline src/Native/System.Native/pal_networking.cpp]: Now that we are using macro,it seems like this conversion is not required (compiling successfully on all three platforms).

**Comment by jasonwilliams200OK** [comment]
I have fixed the issues and nits. PTAL again.

File: src/Native/System.Native/pal_networking.cpp [hunk 6]
```diff
@@ -2321,7 +2336,15 @@ static Error WaitForSocketEventsInner(int32_t port, SocketEvent* buffer, int32_t
...
+#if HAVE_KQUEUE && !HAVE_KEVENT64_S
+            // when kevent64_t is actually kevent
+            // (see the 'typedef kevent' above),
+            // evt.udata is void* and hence static_cast
+            // does not work.
+            .Data = reinterpret_cast<uintptr_t>(evt.udata),
+#else
             .Data = static_cast<uintptr_t>(evt.udata),
+#endif
             .Events = GetSocketEvents(evt.filter, evt.flags)
```
Comment (nguerrera) [inline src/Native/System.Native/pal_networking.cpp]: I'd personally prefer an overloaded GetData(uint64_t) / GetData(void*) than ifdefs. i.e. Prefer letting the compiler reason about types over ifdefs wherever possible.

Comment (nguerrera) [inline src/Native/System.Native/pal_networking.cpp]: @pgavlin with that said, if our data is always pointer sized, why can't we just use the non-64 variant everywhere?

Comment (jasonwilliams200OK) [inline src/Native/System.Native/pal_networking.cpp]: I agree and overloaded method is a pretty decent approach. I just tested and clang is giving me unused function error.

Comment (pgavlin) [inline src/Native/System.Native/pal_networking.cpp]: @nguerrera we could use `kevent` rather than `kevent64`. I originally used `kevent64` for the managed-only implementation due to its fixed size across bitnesses, but that's less of an issue here.

**Comment by pgavlin** [comment]
LGTM.

**Comment by sokket** [comment]
Can you please run clang-format on the CPP and H files that you edited from within the Native directory? :)

File: src/Native/System.Native/pal_networking.cpp (hunk 6 above)
Comment (nguerrera) [inline src/Native/System.Native/pal_networking.cpp]: My vote is to just switch to `kevent` then. It will be cleaner.

Comment (sokket) [inline src/Native/System.Native/pal_networking.cpp]: +1

**Comment by jasonwilliams200OK** [comment]
@sokket, done! :)

**Comment by nguerrera** [comment]
Please keep clang-format runs and actual semantic changes into separate commits.

**Comment by sokket** [comment]
@nguerrera Running clang-format on the specific files that you change is different than the format-code script. After this PR is through I'll do a PR to run the whole script

**Comment by nguerrera** [comment]
@socket, but now the commit is formatting lines that have nothing to do with this change, that should still be in a different commit (doesn't have to be a separate PR, though.)

**Comment by jasonwilliams200OK** [comment]
Agree, I will cherrypick it back and add a new commit. No problemo ;)

**Comment by sokket** [comment]
@nguerrera Ah good point...sorry, I read that as different PR. I agree that it should be a separate commit :)

**Comment by jasonwilliams200OK** [comment]
Now the changes are in separate commits. :)

**Comment by nguerrera** [comment]
Thanks. I'd still like to see this reworked to just use `kevent` instead of `kevent64` everywhere.

File: src/Native/format-code.sh [hunk 7]
```diff
@@ -8,7 +8,7 @@ elif which "clang-format" > /dev/null 2>&1 ; then
    export CF="$(which clang-format)"
 else
    echo "Unable to find clang-format"
-   exit 1
+   export CF="/usr/local/llvm37/bin/clang-format"
```
Comment (sokket) [inline src/Native/format-code.sh]: Why this change?

Comment (jasonwilliams200OK) [inline src/Native/format-code.sh]: Removed. Now the format change is only scoped to the document i have changed in this PR. :)

**Comment by jasonwilliams200OK** [comment]
@nguerrera, should I do the honors and purge kevent64 and co.? :)

**Comment by nguerrera** [comment]
Yes, give it a shot. :)

**Comment by jasonwilliams200OK** [comment]
Done. :tada:

**Comment by pgavlin** [comment]
LGTM.
